### ISSUE TITLE
Classification Facet behavior is not full

### ISSUE THREAD

**Issue by I-Sokolov** [buildingSMART/IDS#223]
1) value: in case of multi-level classification reference - is it refer to top identification or identification at any level?

2) URI - is it refer to IfcClassification.Specification(Location) or to IfcExternalReference.Location

**Comment by andyward** [comment]
1. My understanding is we should check the value at any level in the classification hierarchy (any ancestor). See this [test case](https://github.com/buildingSMART/IDS/blob/development/Documentation/ImplementersDocumentation/TestCases/classification/pass-values_match_subreferences_if_full_classifications_are_used__e_g__ef_25_10_should_match_ef_25_10_25__ef_25_10_30__etc_.ids)
2. URI - This is presumably up to the spec author? But I'd assume a link to the particular classification would be most relevant, rather than the system. It's meta data not a testable attribute I believe

**Comment by I-Sokolov** [comment]
My understanding for 1 is the same, but I'd like to have it clearly declared in the doc.
For 2 it should not be up to spec author because checker developer does not know his intentions:)

**Comment by Moult** [comment]
The URI is not part of the requirement, it's just metadata. Like the URI in other facets.

**Comment by I-Sokolov** [comment]
@Moult, I would disagree, in case of classification facet URI declared as parameter, not metadata

![image](https://github.com/buildingSMART/IDS/assets/6941925/22320430-8231-46e6-a1ad-ef5b0a56ea21)

**Comment by andyward** [comment]
Probably needs clarifying in the docs, but my reading on it is that URI is a parameter you'd suggest be could specified by the author/authoring tool, to aid the ultimate users delivering models against the spec. 

But I don't think there's any expectation it forms part of the verification parameters. At best it might be something take the user to for more information in the event of a verification failure?  

I can also see a structured URI being highly useful to populate a requirement - and indeed some of the object libraries / PDTs services are outputting IDS, where it would make total sense to leave a URI link back to the master data. But there's no saying the URI will be structured. I've seen web pages, PDFs and XLS already in the URI.

**Comment by Moult** [comment]
Yeah I was the one who wrote that table and at the time I believe the intention was to encourage users to use the bSDD where possible to guide users on how to input data: https://github.com/buildingSMART/IDS/commit/4381e77164f120afbfc93bda9c49f83d17e4bb5c - so as you can see it covers classifications, materials, and properties. However the intention was never to actually audit the classification reference location field.

I'll be happy to update the docs to prevent this future confusion, sorry.

**Comment by atomczak** [comment]
> The URI is not part of the requirement, it's just metadata. Like the URI in other facets.

I confirm, the URI is not part of the IDS check, just optional metadata to guide IDS users about the class definition. Documentation needs improvement.

> My understanding is we should check the value at any level in the classification hierarchy (any ancestor). See this [test case](https://github.com/buildingSMART/IDS/blob/development/Documentation/ImplementersDocumentation/TestCases/classification/pass-values_match_subreferences_if_full_classifications_are_used__e_g__ef_25_10_should_match_ef_25_10_25__ef_25_10_30__etc_.ids)

Really?! So according to the [test case](https://github.com/buildingSMART/IDS/blob/development/Documentation/ImplementersDocumentation/TestCases/classification/pass-values_match_subreferences_if_full_classifications_are_used__e_g__ef_25_10_should_match_ef_25_10_25__ef_25_10_30__etc_.ids) if an IDS classification value requires '12' an IfcClassificationReference with a code '1234' will pass right now? In my opinion, it shouldn't. If someone wants to do such a check, they can use regex pattern '12.*'.

**Comment by andyward** [comment]
> >My understanding is we should check the value at any level in the classification hierarchy (any ancestor). See this [test case](https://github.com/buildingSMART/IDS/blob/development/Documentation/ImplementersDocumentation/TestCases/classification/pass-values_match_subreferences_if_full_classifications_are_used__e_g__ef_25_10_should_match_ef_25_10_25__ef_25_10_30__etc_.ids)

> Really?! So [test case](https://github.com/buildingSMART/IDS/blob/development/Documentation/ImplementersDocumentation/TestCases/classification/pass-values_match_subreferences_if_full_classifications_are_used__e_g__ef_25_10_should_match_ef_25_10_25__ef_25_10_30__etc_.ids) if IDS classification value requires '12' an IfcClassificationReference with a code '1234' will pass right now? In my opinion, it shouldn't. If someone wants to do such a check, they can use regex pattern '12.*'.

Agreed it seems overly complex. Those classification tests are a bit confusing, but I believe what it's testing is if the IDS classification value matches a parent/ancestor of IfcClassificationReferences, it should treat it as a pass. Because all classification systems typically employ the parent name as the prefix of the child's name  (i.e. Uniclass' _Pr_60_60_08_33 : Gas-fired boilers_ shares its prefix with its parent _Pr_60_60_08 : Boilers_)  it does feel pretty redundant. I can't think of a public classification system that doesn't do this.

The only other reason I can see for it would be that it does make it slightly simpler for the author - they don't need to apply wildcard:

E.g. "All IfcBoilers must be classified with value **Pr_60_60_08** from Uniclass" - will pick up Gas-fired Boilers and Electric Boilers etc. without needing it to be a pattern. (**If** the ClassificationReferences are exported hierarchically of course)

**Comment by atomczak** [comment]
So this prevents the ability to check that IfcBoilers are exactly Pr_60_60_08 (Boilers) not Gas-fired Boilers and Electric Boilers? With pattern approach we allow both.

**Comment by andyward** [comment]
Yep. And it'd be simpler to implement and explain. Can't say I see many models with full IfcClassificationReference hierarchies anyway so it was always a bit hypothetical.

**Comment by BIMvoice** [comment]
Measured IfcTester's current classification facet matching against the two questions raised here.

**1) URI.** Confirmed by code inspection: `uri` is stored on the facet object but never referenced anywhere in the matching logic, and it is explicitly stripped when the facet appears in `<applicability>`. This matches the docs update already merged into `development` via #381. Worth noting that `ver/1.1.x` still carries the pre-#381 wording, so 1.1 currently has the superseded text.

**2) Value and multi-level matching.** This one is worth clarifying, because the Nov 2024 exchange discussed it as though it were string prefix matching. It is not.

IfcTester walks the explicit `IfcClassificationReference.ReferencedSource` chain and requires an **exact** match of `Identification` against some node in that real, explicitly modelled hierarchy. I tested this directly: a flat reference coded `1234` against an IDS value of `12`, exact, correctly **fails**. There is no accidental substring match. It only "matches a subreference" when the IFC file actually contains a separate, distinct parent `IfcClassificationReference` entity coded `12`.

So the substring collision scenario @atomczak raised does not occur with the current implementation. A regex value of `12.*` against `1234` does pass, which is @andyward's suggested alternative working as intended.

The remaining question is a design one, whether "exact match against an explicitly modelled ancestor" is the behaviour you want to keep, versus exact match only plus regex for anyone wanting prefix semantics. I am not taking a position on that, just reporting the measured behaviour, since the thread looked to be debating a different implementation than the one that ships.

One coverage note while I was in there: the disputed subreference case has no fail companion, so nothing currently proves that a sibling reference at the same level does **not** match. There are 27 classification cases in total, 17 pass and 10 fail.
